lisa/utils.py

```diff
@@ -817,4 +817,81 @@ def __copy__(self):
             top=copy.copy(self.top),
         )
 
+
+def update_wrapper_doc(func, added_by=None, description=None, remove_args=None):
+    """
+    Equivalent to :func:`functools.wraps` that updates the signature by taking
+    into account the wrapper's extra parameters and the given description.
+
+    :param added_by: Add some kind of reference to give a sense of where the
+        new behaviour of the wraps function comes from.
+    :type added_by: str or None
+
+    :param description: Extra description output in the docstring.
+    :type description: str or None
+
+    .. note:: :func:`functools.wraps` is applied by this decorator, which will
+        not work if you applied it yourself.
+    """
+
+    if description:
+        description = '\n{}\n'.format(description)
+
+    remove_args = remove_args if remove_args else set()
+
+    def decorator(f):
+        wrapper_sig = inspect.signature(f)
+        f = functools.wraps(func)(f)
+        f_sig = inspect.signature(f)
+
+        added_params = [
+            desc
+            for name, desc in wrapper_sig.parameters.items()
+            if (
+                name not in f_sig.parameters.keys()
+                # These kinds are usually present in the wrapper to forward
+                # to the wrapped function, so they are not interesting here
+                and not desc.kind in (
+                    inspect.Parameter.VAR_KEYWORD,
```

Comment:
IMO there should be a Big Fat Warning somewhere, at the very least in the doc if not at runtime when we ignore parameters.

I guess the reason you've not gone for a runtime warning is because you have kwargs/pos args in your use of it in the analysis code. If that's the case then I'd be ok with "just" big fat doc warning.

Reply:
This is just the list of parameters added by the decorator that we want to see on the function once decorated. The parameters we are collecting here come from `wrapper_sig`, and we ignore the ones 1) already defined in f() 2) *args and **kwargs since we don't want the decorator's kwargs to be added to the wrapped function signature (that would be wrong, the decorated function has no such parameter unless it already had initially, in which case they are preserved).

```
def decorator(f):
	@update_wrapper_doc(f)
	def wrapper(*args, foo=3, **kwargs):
		if foo == 42:
			return 'bar'
		else:
			return f(*args, **kwargs)
	return wrapper

@decorator
def f(y):
	pass

# The signature of f() should be "y, foo=3" and must not include
# "*args, **kwargs", since these are not parameters of the wrapped function.
```

Reply:
I'm fine with the behaviour, I'd just like to have it explicitly called out to avoid any confusion.